gapic/samplegen/samplegen.py

```diff
@@ -586,12 +619,7 @@ def _validate_write_file(self, body):
             raise InvalidStatement(
                 "Missing key in 'write_file' statement: 'contents'")
 
-        # TODO: check the rest of the elements for valid subfield attribute
-        base = contents_var.split(".")[0]
-        if base not in self.var_defs_:
-            raise UndefinedVariableReference(
-                "Reference to undefined variable: {}".format(base)
-            )
+        self.validate_expression(contents_var)
 
     def _validate_loop(self, body):
```

Comment:
I think the name `body` here is confusing, because you're referring to the whole `loop` directive, but one of its fields is itself called `body`. Maybe you can s/`body`/`loop`/ ? wdyt?

Reply:
I think `s/body/loop` is an improvement.